src/main/java/decodes/tsdb/algo/ToIrregularUsingPattern.java

```diff
@@ -0,0 +1,252 @@
+//AW:INPUTS_END
+
+//AW:LOCALVARS
+	// Enter any local class variables needed by the algorithm.
+	private CTimeSeries inputTS = null;
+	private CTimeSeries patternTS = null;
+	private String inUnits = null;
+//AW:LOCALVARS_END
+
+//AW:OUTPUTS
+	public NamedVariable output = new NamedVariable("output", 0);
+	String _outputNames[] = { "output" };
+//AW:OUTPUTS_END
+
+//AW:PROPERTIES
+	String _propertyNames[] = {};
+//AW:PROPERTIES_END
+
+	// Allow javac to generate a no-args constructor.
+
+	/**
+	 * Algorithm-specific initialization provided by the subclass.
+	 */
+	protected void initAWAlgorithm( )
+		throws DbCompException
+	{
+//AW:INIT
+		_awAlgoType = AWAlgoType.TIME_SLICE;
+//AW:INIT_END
+
+//AW:USERINIT
+		// Code here will be run once, after the algorithm object is created.
+//AW:USERINIT_END
+	}
+	
+	/**
+	 * This method is called once before iterating all time slices.
+	 */
+	protected void beforeTimeSlices()
+		throws DbCompException
+	{
+//AW:BEFORE_TIMESLICES
+		
+		// Validation
+		String outputIntvs = getParmRef("output").compParm.getInterval();
+		Interval outputIntv = IntervalCodes.getInterval(outputIntvs);
+		String patternIntvs = getParmRef("pattern").compParm.getInterval();
+		Interval patternIntv = IntervalCodes.getInterval(patternIntvs);
+
+		if(!outputIntv.equals(patternIntv)){
+			throw new DbCompException("Output interval does not match Pattern interval");
+		}
+
+		inputTS = getParmRef("input").timeSeries;
+		patternTS = getParmRef("pattern").timeSeries;
+
+		inUnits = getInputUnitsAbbr("input");
+		setOutputUnitsAbbr("output", inUnits);
+
+		//Check for any uninterpolated  pattern dates in the past and interpolate with new inputs
+		Date firstInputT = baseTimes.first();
+		try {
+			TimedVariable prevInput = tsdb.getPreviousValue(inputTS, firstInputT);
+			TimedVariable nextInput = tsdb.getNextValue(inputTS, firstInputT);
+			boolean lookback = true;
+			TimedVariable prevPattern = tsdb.getPreviousValue(patternTS, firstInputT);
+			while(lookback){
+				if (0 > prevInput.getTime().compareTo(prevPattern.getTime())){
+					lookback = false;
+				}
+				else{
+					interpolate(prevPattern.getTime(), prevInput, nextInput);
+					prevPattern = tsdb.getPreviousValue(patternTS, prevPattern.getTime());
+				}
+			}
+
+		}
+		catch(Exception ex)
+		{
+			warning("Error accessing input/output time series: " + ex);
+		}
+
+		
+//AW:BEFORE_TIMESLICES_END
+	}
+
+	/**Interpolation helper function
+	 * Date Pattern data - data time which you want to interpolate the value for.
+	 * TimedVariable prev - TimedVariable of the nearest in time previous input variable
+	 * TimedVariable next - TimedVariable of the nearest in time next input variable
+	 * Checks the type of data stored in the input and preforms appropriated interpolation between two points.
+	**/
+	void interpolate(Date patternDate, TimedVariable prev, TimedVariable next){
+		CwmsTsId tsId = (CwmsTsId) inputTS.getTimeSeriesIdentifier();
```

Comment:
Remind me why this is currently CWMS specific. Did the TimeSeriesIdentifier not provide sufficient information?

Reply:
only CwmsTsId have the variable ParamType (Inst, Total, Ave) which is needed to check what type of interpolation we need to preform. Do you recommend I add a check to the algorithm to see if the input is from Cwms and if it is not use default linear Interpolation.

Reply:
Ok I will make sure to do that.

Reply:
@agilmore2 does HDB provide equivalent information in some way?
e.g.  Is the data already an average, instantatenous, a total between interverals, etc, but just using different names in the code?